# Assignment 5: Association mining

## Objective of this assignment
The overall objective is to understand how frequent itemsets can be extracted by
the Apriori algorithm and be able to calculate and interpret association rules in terms of support and confidence.

## ** Important: ** When handing in your homework:
+ Hand in the notebook (and nothing else) named as follows: StudentName1_snumber_StudentName2_snumber.ipynb
+ Provide clear and complete answers to the questions below under a separate header (not hidden somewhere in your source code), and make sure to explain your answers / motivate your choices. Add Markdown cells where necessary.
+ Source code, output graphs, derivations, etc., should be included in the notebook.
+ Hand-in: upload to Blackboard.
+ Include name, student number, assignment (especially in filenames)!
+ When working in pairs only one of you should upload the assignment, and report the name of your partner in your filename.
+ For problems or questions: use the BB discussion board or email the student assistants.


## Advised Reading and Exercise Material
**The following reading material is recommended:**

- Pang-Ning Tan, Michael Steinbach, and Vipin Kumar, *Introduction to Data Mining*, section 6.


## Additional Tools
For this exercise you will need to load the provided *apriorimining.py* script. 


##  5.1 Association mining for course data 
We will use the Apriori algorithm to automatically mine for associations. The Apriori algorithm is an adapted version of the script found here: https://github.com/nalinaksh/Association-Rule-Mining-Python

Check out the script and doc and check if you understand how the association rules are computed. 



In [3]:
import Toolbox.apriorimining as apriorimining
# Script and doc checked


#### 5.1.1

(0 points) Look at the data file `Data/courses.txt` into Python. The data is represented in Table 1. Inspect the file Data/courses.txt and make sure you understand how the data in Table 1 is stored in the text file.

##### Table 1
|#  |   History |Math| Biology| Spanish | Economics| Physics | Chemistry | English  |  
| :-------------: |:-------------:| :-----------:| :----------:| :------------:|:-------------:| :------------:|  :-------------: | :-------------: |
|student 1 | 0| 1 | 0 | 0 | 1| 1 |1 |1   
|student 2 | 1| 1 | 1 | 0 | 0| 1 |1 |1   
|student 3 | 0| 1 | 0 | 1 | 0| 1 |0 |1   
|student 4 | 0| 0 | 1 | 0 | 0| 1 |1 |0   
|student 5 | 0| 1 | 0 | 0 | 0| 1 |1 |0        
|student 6 | 0| 1 | 1 | 0 | 0| 1 |1 |1   


#### 5.1.2
(1 point) We will analyze the data in Table 1 automatically using the function `apriorimining.generate_association_rules()` from the script. Analyze the data with $ minsupport  \geq 80 \% $ and $ minconfidence \geq 100 \%$.What
are the generated association rules? What kind of conclusions can you make based on these association rules about the subjects that students took?  
  


In [5]:
apriorimining.generate_association_rules()

Please enter support value in %: 80
Please enter confidence value in %: 100
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename (for Windows), or filepath/filename (for UNIX/Mac), and extension: /Users/tristan/Documents/Data Mining/Assignment_5/Data/courses.txt
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { 6 },  sup= 100.0
set= { 2 },  sup= 83.33
set= { 7 },  sup= 83.33
-----------------------------------------------------------------
-------TOP 10 (or less) FREQUENT 2-ITEMSET------------------------
set= { 2, 6 },  sup= 83.33
set= { 6, 7 },  sup= 83.33
------------------------------------------------------------------
---------------------ASSOCIATION RULES------------------
--------------------------------------------------------
Rule #1: {  } ==> { 6 }, sup= 100.00, conf= 100.00

Rule #2: { 2 } ==> { 6 }, sup= 83.33, conf= 100.00

Rule #3: { 7 } ==> { 6 }, sup= 83.33, conf= 100.00
----------------------

1. Every student follows Physics
2. Students following Math also follow Physics (although this is already clear from rule 1)
3. Students following Chemistry also follow Physics (although this is already clear from rule 1)

  ##  5.2 Association mining for MovieLens data 
  
  
  In this part of the exercise we consider a Market Basket data set containing 943 users purchases of 1682 movies. A total of 100,000 movies
have been purchased.The data set is called MovieLens100K and is provided by http://www.grouplens.org/node/73, see also the readme `MovieLensData.txt` in the data folder. The data currently considered is not the original data but modified for the apriori algorithm.

#### 5.2.1
  (0 points) The MovieLens data is stored in the file MovieLensData.txt. Inspect the file to see how the data is stored.


#### 5.2.2 
  (1 point) Find association rules using the function below with $ minsupport  \geq 30 \% $ and $ minconfidence \geq 80 \%$. What are the associations with strongest confidence? Do these associations make sense? The script can use file Data/u.item to print the movie titles in stead of numbers. If you enter filename `MovieLensData.txt`, the script will provide an additional option for this. 
  

In [10]:
apriorimining.generate_association_rules()

Please enter support value in %: 30
Please enter confidence value in %: 80
Enter the max number of rules you want to see (enter 0 to see all rules): 0
Please enter filepath\filename (for Windows), or filepath/filename (for UNIX/Mac), and extension: /Users/tristan/Documents/Data Mining/Assignment_5/Data/MovieLensData.txt
Do you want to print sets and rules with Movie names in stead of numbers? [y/n]: i guess that would be nice
---------------TOP 10 FREQUENT 1-ITEMSET-------------------------
set= { Star Wars (1977) },  sup= 61.82
set= { Contact (1997) },  sup= 53.98
set= { Fargo (1996) },  sup= 53.87
set= { Return of the Jedi (1983) },  sup= 53.76
set= { Liar Liar (1997) },  sup= 51.43
set= { English Patient, The (1996) },  sup= 51.01
set= { Scream (1996) },  sup= 50.69
set= { Toy Story (1995) },  sup= 47.93
set= { Air Force One (1997) },  sup= 45.71
set= { Independence Day (ID4) (1996) },  sup= 45.49
-----------------------------------------------------------------
-------TOP 10 (or le

Many of the rules with the highest confidence relate to the original star wars movies. This is a series, and so it makes sense that if you have seen one of the movies, you will have seen the others! Raiders of the lost ark is associated with the star wars movies as well. This makes sense if you consider that it was also written by George Lucas, and it also has Harrison Ford.

#### 5.2.3 
(1 point) Which movies have been watched by the most users? There are only few rules with more than three items. Why?

The three most watched movies are Star wars (1977), Contact, and Fargo.   (which are nice movies!)


The reason that there are only few rules with more than three items is, i believe, related to our choice of the minimum support level. It appears that it is not common for the same four movies to each be present in more than 30% of the dataset (We only find one such instance from the support settings indicated by the question). This means that there is not a lot of room to generate rules from this single set.

#### 5.2.4
(0.5 points) Often we are interested in rules with high confidence. Is it possible for
itemsets to have very low support but still have a very high confidence?

Yes, if we set minimum support lower, the rules will generally get more freedom because there are more combinations available to make rules from. E.g. if we set minimum support to 0%, the algorithm can make perfect, but complex rules at the single-object level. However, these rules will not generalize at all and won't be very meaningful. I suppose, in some ways, this might be thought of as some form of over-fitting.

## 5.3 Calculating support, confidence and interest

Calculate these measures and write down how you computed things, not just the answers. 


#### 5.3.1
 Suppose we have market basket data consisting of 100 transactions and 20 items. The support for item $ \text{a} = 45 \%$, the support for item $ \text{b} = 80 \%$ and the support for itemset $ \text{ {a,b }} = 30 \%$. Let the support and confidence thresholds be 20$ \%$ and 60$ \%$, respectively.
  
1. (0.5 points) Compute the confidence of the association rule $ \text{ {a } } \rightarrow   \text{{b }} $. Is the rule interesting according to the confidence measure?

2. (0.5 points) Compute the interest measure (or lift, see slide 44 of chapter 6) for the association pattern $ \text{ {a,b}}$. Describe the nature of the relationship between item $ \text{a}$ and item $ \text{b}$  in terms of the itemset measure.
3. (1 points) What conclusion can you draw from the results of parts (1) and (2)?

QUESTION 1

To assist in computing this, i created a contingency table (on paper) similar to lecture slide 42.
P(B|A) = P(A&B)/P(A) = 0.3/0.45 = 2/3
This rule is not as interesting as the items appear to be negatively correlated (stated by prof. Heskes in the lecture). This is because in our example P(A&B) = 0.3, whereas if they had been independent we would expect P(A&B) to be P(A)P(B) = 0.36. Our value is lower -> They are negatively correlated and not as interesting to us,  it would have been more interesting if they had been positively correlated instead. We do surpass the confidence threshold of 60%, as the confidence for this rule is 66%. We also surpass the support threshold of 20% as itemset (A,B) has a support of 30%.


QUESTION 2

Interest is P(A,B)/P(A)P(B) = 0.3/0.36 = 0.833. This is lower than 1, and as such implies a negatively correlated relationship between item A and item B.


QUESTION 3

We now know that customers that buy product A are less likely to also buy product B than customers that did not buy product A. The rule is therefore somewhat misleading (as it is suggested that a would predict b, and this passes the confidence threshold). a does predict b, but given the base rates for a and b, it predicts it less than we would expect by chance. This rule is therefore not very useful.





#### 5.3.2

1. (1 points) Let $c_1$, $c_2$, and $c_3$ be the confidence values of the rules $ \text{ {a } } \rightarrow   \text{{b }} $, $ \text{ {a } } \rightarrow   \text{{b,c }} $, and $ \text{ {a,c } } \rightarrow   \text{{b }} $ respectively. If we assume that $c_1$, $c_2$, and $c_3$ have different values, what are the possible inequality relationships (e.g. $c_1 \leq c_2 \leq c_3$) among $c_1$, $c_2$, and $c_3$? Which rule has the lowest confidence?
2. (0.5 points) Suppose the confidence of the rules  $ \text{ {a } } \rightarrow   \text{{b }} $ and $ \text{ {b } } \rightarrow   \text{{c }} $ are larger than the confidence threshold. Is it possible that $ \text{ {a } } \rightarrow   \text{{c }} $ has a confidence below that threshold? If no, explain why. If yes, give an example. 

QUESTION 1

From theorem 6.2 in the book we know that if rule X -> Y - X, it is the case that for any subset X' of X, it is the case that X' -> Y - X' must necessarily have a lower (or equal) confidence. Now we can first of all assert that it must be the case that c2 $\leq$  c3. for c1, this is not as trivial as it is not the same set as c2 and c3. However, as confidence is computed as the support of the whole set divided by the support of the antecedent, and c1, c2 have the same antecedent, and we know from the antimonotonicity property of itemset support that support for {a,b} must be greater or equal to support for {a,b,c}. This means that support{a,b}/support{a} will be greater or equal to support{a,b,c}/support{a}. This means that confidence c2 $\leq$ c1. 

There now still are two possible relationships between the three values: c2 $\leq$ c3 $\leq$ c1 or c2 $\leq$ c1 $\leq$ c3. We can however be certain that c2 has the lowest confidence of the three.



QUESTION 2

This is certainly possible. Take, for instance the following dataset:

1. a,b
2. b,c
3. a,b
4. a,b
5. a,b
6. b,c
7. a,c
8. b,c

The confidence for rule a -> b is 4/5 = 80% 
The confidence for rule b -> c is 3/3 = 100%
However, the confidence for rule a -> c is only 1/5 = 20%

Therefore we see that while rule a -> b and b -> c may be above some threshold (in this example, it could be anywhere between 21% and 80%), rule a -> c can be below it.


#### 5.3.3

(3 points) Consider the relationships between customers who buy high-definition televisions and exercise machines as shown in Table 2 and 3.

1. Compute the odd ratios for both tables.
2. Compute the $\phi$-coefficient for both tables.
3. Compute the interest (or lift, in the book) factor for both tables.

For Table 3 you should compute measures given above separately for College
Students and for Adults. For each of the measures, describe how the direction
of association changes when data is pooled together (Table 2) instead of being
separated into two groups (Table 3)

##### Table 2: Two way contingency table between the sale of high-definition television and exercise machine
| |   Buy Exercise machine |     |     |
| :------------- | -------------:| :-----------:| :----------:| 
| **Buy HDTV     ** | yes | no | total |
| yes  | 105| 87 | 192 | 
| no | 40| 62 | 102 |   
| total | 145 | 149 | 294 | 
 

##### Table 3: Example of three-way contingency table
| | |   Buy Exercise machine |     |     |
|--- | :------------- | -------------:| :-----------:| :----------:| 
|**Customer group** | **Buy HDTV     ** | yes | no | total |
|College students | yes  | 2| 9 | 11 | 
| | no | 5| 20 | 25 |
| Working adults | yes  | 103| 78 | 181 | 
| | no | 35| 42 | 77 |  




QUESTION 1 (odds ratio)

Table 2: (105 x 62)/(87 x 40)                                   = 1.871

Table 3a:(2 x 20)/(9 x 5)                                       = 0.888

Table 3b: (103 x 42)/(78 x 35)                                  = 1.585




QUESTION 2 (correlation)

Table 2:  294 x 105 - 192 x 145 / sqrt(192 x 145 x 149 x 102)   = 0.147

Table 3a: 36 x 2 - 11 x 7 / sqrt(11 x 7 x 29 x 25)              = -0.021

Table 3b: 258 x 103 - 181 x 138 / sqrt(181 x 138 x 120 x 77)    = 0.105




QUESTION 3 (interest)

Table 2: (105 x 294)/(192 x 145)                                = 1.109

Table 3a:(2 x 36)/(11 x 7)                                      = 0.935

Table 3b: (103 x 258)/(181 x 138)                               = 1.064



For each of the measures used, the direction of the association in the pooled data is positive, and is increased in amplitude compared to the separated data (i.e. it is not their 'mean'). Notably, the direction association in college students alone even becomes negative. So the association between buying an HDTV and an exercise machine in the pooled data is stronger/more positive than in either of the subgroups separately. This seems somewhat strange but reflects Simpson's paradox where the interpretation of association between variables can be influenced by hidden variables. Interpretation can be improved by stratifying the data in a meaningful way, although it is perhaps not so trivial to determine which stratificaton is 'meaningful'.
